bus-mapping/src/evm/opcodes/extcodehash.rs

```diff
@@ -0,0 +1,216 @@
+use super::Opcode;
+use crate::circuit_input_builder::CircuitInputStateRef;
+use crate::operation::{
+    AccountField, AccountOp, CallContextField, CallContextOp, TxAccessListAccountOp, RW,
+};
+use crate::Error;
+use eth_types::{evm_types::GasCost, GethExecStep, ToAddress, ToWord};
+
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Extcodehash;
+
+impl Opcode for Extcodehash {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        steps: &[GethExecStep],
+    ) -> Result<(), Error> {
+        let step = &steps[0];
+        let stack_address = step.stack.last_filled();
+
+        // Pop external address off stack
+        let external_address = step.stack.last()?.to_address();
+        state.push_stack_op(RW::READ, stack_address, external_address.to_word());
+
+        // Read transaction id from call context
+        state.push_op(
+            RW::READ,
+            CallContextOp {
+                call_id: state.call().call_id,
+                field: CallContextField::TxId,
+                value: state.tx_ctx.id().into(),
+            },
+        );
+
+        // Update transaction access list for external_address
+        let account_previously_accessed = match step.gas_cost {
+            GasCost::WARM_STORAGE_READ_COST => true,
+            GasCost::COLD_ACCOUNT_ACCESS_COST => false,
+            _ => unreachable!(),
+        };
+        state.push_op(
+            RW::WRITE,
+            TxAccessListAccountOp {
+                tx_id: state.tx_ctx.id(),
+                address: external_address,
+                value: true,
+                value_prev: account_previously_accessed,
+            },
+        );
+
+        // Account read for code hash of external account
+        let code_hash = steps[1].stack.last()?;
+        state.push_op(
+            RW::READ,
+            AccountOp {
+                address: external_address,
+                field: AccountField::CodeHash,
+                value: code_hash,
+                value_prev: code_hash,
+            },
+        );
+
+        // Stack write of the the code
+        state.push_stack_op(RW::WRITE, stack_address.map(|a| a - 1), code_hash);
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod extcodehash_tests {
```

Comment:
From running these tests and printing the traces I've observed that the gas cost for non-existing accounts is `WARM_STORAGE_READ_COST`.  The way this is structured causes that for the test `cold_empty_account`, the following op is generated:
```
TxAccessListAccountOp { tx_id: 1, addr: 0x0000000000000000000000000000000000000000, val_prev: true, val: true }
```
Which is not correct because in a Transaction the value for `TxAccessListAccountOp` should always start at false.  This condition should be enforced by the state circuit, so even though the EVM circuit passes the tests, the `rw_table` generated by the circuit input builder will not pass the state circuit constraints.

So I'm wondering: maybe for non-existing accounts we shouldn't generate the `TxAccessListAccountOp`?  As the concept of cold / warm doesn't apply, the gas cost for non-existing account is constant.  Since the gas cost for non-existing accounts will always be the same, the circuit needs to prove that the account doesn't exist.  That condition is true when `Nonce = 0, Balance = 0 and CodeHash = keccack([])`.  So maybe we need to do these 3 lookups to detect if the account exists or not.

Reply:
The `cold_existing_account` test was not actually doing what it said, because the executing address is never cold, because it starts in the accessed accounts list. I believe the `TxAccessListAccountOp` that you posted above it from that test, because the `addr` is of the default account, which is what the tests use to execute the byte code.

Switching to using the `is_warm` variable in the test actually caught this.

Reply:
The yellow paper defines empty accounts as ones with nonce, balance, and code all at their default values. It's not enough to check just the nonce, because e.g. you could send eth to a previously empty account, which would not increase its nonce.

I think it would be sound to only verify that nonce = 0 and balance = 0, because https://eips.ethereum.org/EIPS/eip-161 requires incrementing the nonce of an account before deploying a contract there, but I figure we can remove the code hash check in the future if we need to make that optimization.

Reply:
> The `cold_existing_account` test was not actually doing what it said, because the executing address is never cold, because it starts in the accessed accounts list. I believe the `TxAccessListAccountOp` that you posted above it from that test, because the `addr` is of the default account, which is what the tests use to execute the byte code.

You're right!  The `TxAccessListAccountOp` I wrote was from the `cold_existing_account` test (I mistakenly thought it was from the `cold_empty_account` test).  It makes sense that the account is already warm as it's used in the transaction.

That mistake made me infer the wrong conclusions.  So after re-evaluating: the concept of cold / warm still applies to non-existing accounts, and they have different gas price for cold / warm access.

We still need to test for non-existing accounts, just to prove when the opcode result is 0.

Also you're right, checking the nonce alone is not enough!  Checking nonce, balance and code hash sounds good at this point :)